### PR TITLE
[tests] add traffic analysis for Cert_5_2_01 as TestPlan

### PR DESCRIPTION
**Pull request by zhaxia**
This commit also modified the method assertSentToNode to confirm if the destination Mac address of the message is the node's Mac address.

### CODE REVIEW

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+        self.assertEqual(self.nodes[DUT_ROUTER1].get_state(), 'router')
+
+        # DUT_ROUTER1: Verify MLE advertisements
+        router1_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+        msg = router1_messages.next_mle_message(mle.CommandType.ADVERTISEMENT)
+        msg.assertSentWithHopLimit(255)
+        msg.assertSentToDestinationAddress('ff02::1')
+        msg.assertMleMessageContainsTlv(mle.SourceAddress)
+        msg.assertMleMessageContainsTlv(mle.LeaderData)
+        msg.assertMleMessageContainsTlv(mle.Route64)
```
Comment (hjian2017) [inline tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: There is a method(check_mle_advertisement) in command.py.  You can reuse it here.

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        msg.assertCoapMessageRequestUriPath('/a/as')
+
+        # DUT_ROUTER1: Verify Address Solicit Response to REED1
+        msg = router1_messages.next_coap_message('2.04')
+        msg.assertSentToDestinationAddress(reed1_ipv6_address)
+
+        # 9 LEADER: Verify connectivity by sending an ICMPv6 Echo Request to REED1 
+        for addr in self.nodes[REED1].get_addrs():
+            if addr[0:4] != 'fe80':
+                self.assertTrue(self.nodes[LEADER].ping(addr))
```
Comment (hjian2017) [inline tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: We have a method(get_ip6_address) in node_cli.py. You can get any address by this method.

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+LEADER = 1
+DUT_ROUTER1 = 2
+REED1 = 3
+ED1 = 4
+
+class Cert_5_2_01_REEDAttach(unittest.TestCase):
+    def setUp(self):
+        self.nodes = {}
+        for i in range(1, 5):
+            self.nodes[i] = node.Node(i)
```
[the remark this answers is not in the document]
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Has updated

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        del self.nodes
+
+    def test(self):
+        self.nodes[LEADER].start()
+        self.nodes[LEADER].set_state('leader')
+        self.assertEqual(self.nodes[LEADER].get_state(), 'leader')
+
+        # 1 DUT_ROUTER1: Attach to LEADER
+        self.nodes[DUT_ROUTER1].start()
+        time.sleep(10)
```
[the remark this answers is not in the document]
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Has updated

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        msg.assertMleMessageContainsTlv(mle.LeaderData)
+        msg.assertMleMessageContainsTlv(mle.NetworkData)
+        msg.assertMleMessageContainsTlv(mle.SourceAddress)
+        msg.assertMleMessageContainsOptionalTlv(mle.Route64)
+
+        # 5 Omitted
+
+        # 6 ED1: Attach to REED1
+        self.nodes[ED1].start()
+        time.sleep(20)
```
[the remark this answers is not in the document]
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Has updated

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+        self.assertEqual(self.nodes[ED1].get_state(), 'child')
```
[the remark this answers is not in the document]
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: This test case is focus on testing the DUT_ROUTER1. I don't think we need to check the state of REED1.

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+
+        # 7 REED1: Verify Address Solicit Request to DUT_ROUTER1
+        reed1_messages = self.sniffer.get_messages_sent_by(REED1)
+        msg = reed1_messages.next_coap_message('0.02')
+        reed1_ipv6_address = msg.ipv6_packet.ipv6_header.source_address.compressed
+        msg.assertSentToNode(self.nodes[DUT_ROUTER1]);
+        msg.assertCoapMessageRequestUriPath('/a/as')
+
+        # 8 DUT_ROUTER1: Verify Address Solicit Requests to LEADER
```
[the remark this answers is not in the document]
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Has Updated.

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        router1_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+        msg = router1_messages.next_coap_message('0.02')
+        msg.assertSentToNode(self.nodes[LEADER]);
+        msg.assertCoapMessageRequestUriPath('/a/as')
+
+        # DUT_ROUTER1: Verify Address Solicit Response to REED1
```
[the remark this answers is not in the document]
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Has Updated.

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+        msg = router1_messages.next_coap_message('2.04')
+        msg.assertSentToDestinationAddress(reed1_ipv6_address)
```
[the remark this answers is not in the document]
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: The Address Solicit Response message doesn't contain an "a/as" uri path. So we don't add "a/as" uri path verification.

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py (hunk 1 above)
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Good idea. These commands have been encapsulated in command.py.

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (librasungirl) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: `self.nodes[i] = node.Node(i, i == MED1)`

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        self.assertEqual(self.nodes[MED1].get_state(), 'child')
+
+        # 7 REED1: Verify sending Address Solicit Request to DUT_ROUTER1
+        reed1_messages = self.sniffer.get_messages_sent_by(REED1)
+        msg = reed1_messages.next_coap_message('0.02')
+        reed1_ipv6_address = msg.ipv6_packet.ipv6_header.source_address.compressed
+        msg.assertSentToNode(self.nodes[DUT_ROUTER1]);
+        msg.assertCoapMessageRequestUriPath('/a/as')
+
+        # 8 DUT_ROUTER1: Verify forwarding REED1's Address Solicit Requests to LEADER
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: nit: Requests->Request

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/env python
+#
+#  Copyright (c) 2016, The OpenThread Authors.
```
Comment (librasungirl) [inline tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: 2016->2017?

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py (hunk 10 above)
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Has Updated

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py (hunk 11 above)
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Has Updated

**Comment by hjian2017** [comment]
LGTM

File: tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+        router1_messages = self.sniffer.get_messages_sent_by(DUT_ROUTER1)
+        msg = router1_messages.next_mle_message(mle.CommandType.ADVERTISEMENT)
+        command.check_mle_advertisement(msg)
+
+        # 2 REED1: Attach to DUT_ROUTER1
+        self.nodes[REED1].start()
+        time.sleep(5)
+        self.assertEqual(self.nodes[REED1].get_state(), 'child')
+
+        # 3 DUT_ROUTER1: Verify MLE Rarent Response
```
[the remark this answers is not in the document]
Comment (zhaxia) [reply tests/scripts/thread-cert/Cert_5_2_01_REEDAttach.py]: Has updated

File: tests/scripts/thread-cert/Makefile.am [hunk 13]
```diff
@@ -174,7 +174,7 @@ check_SCRIPTS                                                      = \
     Cert_5_1_11_REEDAttachLinkQuality.py                             \
     Cert_5_1_12_NewRouterNeighborSync.py                             \
     Cert_5_1_13_RouterReset.py                                       \
-    Cert_5_2_01_BecomeActiveRouter.py                                \
+    Cert_5_2_01_REEDAttach.py                                \
```
Comment (jwhui) [inline tests/scripts/thread-cert/Makefile.am]: nit: align `\`

Comment (zhaxia) [reply tests/scripts/thread-cert/Makefile.am]: Has updated

**Pull request merged**